### PR TITLE
feat(video): add support for contour ROI and contour based segmentation

### PR DESCRIPTION
**Pull request by wayfarer3130**
### Context
Add spline rendering on video viewports

### Testing

Try it https://deploy-preview-988--cornerstone-3d-docs.netlify.app/live-examples/videocontoursegmentation

#### PR

- [] My Pull Request title is descriptive, accurate and follows the
  semantic-release format and guidelines.

#### Code

- [] My code has been well-documented (function documentation, inline comments,
  etc.)

- [] I have run the `yarn build:update-api` to update the API documentation, and have
  committed the changes to this PR. (Read more here https://www.cornerstonejs.org/docs/contribute/update-api)

#### Public Documentation Updates

- [] The documentation page has been updated as necessary for any public API
  additions or removals.

#### Tested Environment

- [] "OS: 
- [] "Node version: 
- [] "Browser:
  

[blog]: https://circleci.com/blog/triggering-trusted-ci-jobs-on-untrusted-forks/
[script]: https://github.com/jklukas/git-push-fork-to-upstream-branch

### CODE REVIEW

File: utils/demo/helpers/addNavigationBindings.ts [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+      },
+    ],
+  });
+  toolGroup.setToolActive(LengthTool.toolName, {
+    bindings: [
+      {
+        mouseButton: MouseBindings.Fourth_Button,
+      },
+    ],
+  });
```
Comment (sedghi) [inline utils/demo/helpers/addNavigationBindings.ts:68]: This is odd to have activated always

File: utils/ExampleRunner/example-info.json [hunk 2]
```diff
@@ -326,6 +326,10 @@
         "name": "Spline ROI Tools",
         "description": "Demonstrates how to use spline ROI tools (Linear, Cardinal, Catmull-ROM and BSpline)"
       },
+      "videoSplineROITools": {
```
Comment (sedghi) [inline utils/ExampleRunner/example-info.json:329]: move to video section please

File: packages/core/src/RenderingEngine/VideoViewport.ts [hunk 3]
```diff
@@ -216,10 +220,11 @@ class VideoViewport extends Viewport implements IVideoViewport {
...
       // 1 based range setting
       this.setFrameRange([1, numberOfFrames]);
-      if (frameNumber !== undefined) {
+      this.play();
+      window.setTimeout(() => {
         this.pause();
-        this.setFrameNumber(frameNumber);
-      }
+        this.setFrameNumber(frameNumber || 1);
+      }, 50);
```
Comment (sedghi) [inline packages/core/src/RenderingEngine/VideoViewport.ts:227]: requires a comment explaining what is this 50 ms timeout

File: packages/core/src/RenderingEngine/VideoViewport.ts [hunk 4]
```diff
@@ -259,16 +264,24 @@ class VideoViewport extends Viewport implements IVideoViewport {
...
-      this.isPlaying = true;
-      this.renderWhilstPlaying();
+    try {
+      if (!this.isPlaying) {
+        this.videoElement.play();
+        this.isPlaying = true;
+        this.renderWhilstPlaying();
+      }
+    } catch (e) {
+      // No-op, this happens sometimes on startup
```
Comment (sedghi) [inline packages/core/src/RenderingEngine/VideoViewport.ts:274]: does this "this happens sometimes on startup" mean sometimes the auto-play won't work?

File: packages/tools/examples/splineROITools/index.ts [hunk 5]
```diff
@@ -273,22 +266,7 @@ async function run() {
...
-  });
-
-  toolGroup.setToolActive(ZoomTool.toolName, {
-    bindings: [
-      {
-        mouseButton: MouseBindings.Secondary, // Right Click
-      },
-    ],
-  });
+  addNavigationBindings(toolGroup);
```
Comment (sedghi) [inline packages/tools/examples/splineROITools/index.ts:269]: I was thinking we call pan, zoom, scroll etc `manipulatoin` in the code base elsewhere and not navigation -> `addManipulationBindings`

File: packages/tools/src/tools/annotation/LivewireContourTool.ts [hunk 6]
```diff
@@ -111,20 +114,19 @@ class LivewireContourTool extends AnnotationTool {
     const FrameOfReferenceUID = viewport.getFrameOfReferenceUID();
     const defaultActor = viewport.getDefaultActor();
 
-    if (!defaultActor || !csUtils.isImageActor(defaultActor)) {
-      throw new Error('Default actor must be an image actor');
-    }
+    // if (!defaultActor || !csUtils.isImageActor(defaultActor)) {
+    //   throw new Error('Default actor must be an image actor');
+    // }
```
Comment (sedghi) [inline packages/tools/src/tools/annotation/LivewireContourTool.ts:119]: not sure why you are deleting this, since even stack viewport is imageActor

File: packages/tools/src/utilities/convertToGrayscale.ts [hunk 7]
```diff
@@ -0,0 +1,27 @@
+export default function convertToGrayscale(
```
Comment (sedghi) [inline packages/tools/src/utilities/convertToGrayscale.ts:1]: can you put this inside core maybe as utilitis?

File: packages/tools/src/utilities/livewire/LivewireScissors.ts [hunk 8]
```diff
@@ -4,6 +4,14 @@ import { BucketQueue } from '../BucketQueue';
 const MAX_UINT32 = 4294967295;
 const TWO_THIRD_PI = 2 / (3 * Math.PI);
 
+type PixelDataTypedArray =
```
Comment (sedghi) [inline packages/tools/src/utilities/livewire/LivewireScissors.ts:7]: we have a type Types.PixelDataTypedArray

File: packages/tools/examples/videoSplineROITools/index.ts [hunk 9]
```diff
@@ -0,0 +1,318 @@
...
+    const splineConfig = toolGroup.getToolConfiguration(
+      splineToolName,
+      'spline'
+    );
+
+    splineConfig.configuration.CARDINAL.scale = value;
+    toolGroup.setToolConfiguration(splineToolName, { spline: splineConfig });
+  },
+});
+
```
Comment (sedghi) [inline packages/tools/examples/videoSplineROITools/index.ts:218]: I think these are not necessary for the sake of this example, user can explore the spline settings in the spline example, or maybe use one only

**Review by sedghi — COMMENTED** [review]
see my comments please

File: packages/tools/src/tools/annotation/LivewireContourTool.ts (hunk 6 above)
Comment (wayfarer3130) [reply packages/tools/src/tools/annotation/LivewireContourTool.ts:119]: It throws on VideoViewport as that doesn't have a default actor (yet).

File: packages/tools/examples/videoSplineROITools/index.ts (hunk 9 above)
Comment (wayfarer3130) [reply packages/tools/examples/videoSplineROITools/index.ts:218]: I could also add the spline tool to the videoTools example, and just let it be there, without further embellishment?

**Review by sedghi — APPROVED** [review]

File: packages/tools/examples/livewireContourSegmentation/index.ts [hunk 10]
```diff
@@ -64,16 +62,16 @@ setTitleAndDescription(
 );
 
 const content = document.getElementById('content');
-const viewoprtsContainer = document.createElement('div');
+const viewportsContainer = document.createElement('div');
```
Comment (sedghi) [inline packages/tools/examples/livewireContourSegmentation/index.ts:65]: Thanks

File: utils/demo/helpers/addManipulationBindings.ts [hunk 11]
```diff
@@ -0,0 +1,70 @@
+import * as cornerstoneTools from '@cornerstonejs/tools';
```
Comment (sedghi) [inline utils/demo/helpers/addManipulationBindings.ts:1]: I guess you need to git pull origin

File: packages/tools/examples/videoSplineROITools/index.ts [hunk 12]
```diff
@@ -0,0 +1,318 @@
...
+    StudyInstanceUID: '2.25.96975534054447904995905761963464388233',
+    SeriesInstanceUID: '2.25.15054212212536476297201250326674987992',
+    wadoRsRoot:
+      getLocalUrl() || 'https://d33do7qe4w26qo.cloudfront.net/dicomweb',
+  });
+
+  // Only one SOP instances is DICOM, so find it
+  const videoId = imageIds.find(
+    (it) => it.indexOf('2.25.179478223177027022014772769075050874231') !== -1
+  );
```
Comment (sedghi) [inline packages/tools/examples/videoSplineROITools/index.ts:288]: Can we please add a section to here https://www.cornerstonejs.org/docs/concepts/cornerstone-core/viewports to explain we have a video viewport and how to set an actual video there?

And ideally a new tutorial https://www.cornerstonejs.org/docs/category/tutorials

We have budget left :)

File: packages/tools/examples/videoContourSegmentation/index.ts [hunk 13]
```diff
@@ -0,0 +1,484 @@
...
+
+  element.addEventListener(Enums.Events.IMAGE_RENDERED, (evt: any) => {
+    const { time, duration } = evt.detail;
+    rangeElement.value = time;
+    rangeElement.max = duration;
+    const timeElement = document.getElementById('time');
+    timeElement.innerText = seconds(time);
+    const remainingElement = document.getElementById('remaining');
+    remainingElement.innerText = seconds(duration - time);
+  });
```
Comment (sedghi) [inline packages/tools/examples/videoContourSegmentation/index.ts:444]: can you refactor this logic outside, it is out of place IMO

File: packages/tools/examples/videoContourSegmentation/index.ts [hunk 14]
```diff
@@ -0,0 +1,484 @@
...
+  // Add a segmentation that will contains the contour annotations
+  segmentation.addSegmentations([
+    {
+      segmentationId,
+      representation: {
+        type: csToolsEnums.SegmentationRepresentations.Contour,
+        data: {
+          // geometryIds may not be used anymore because it will be removed in a
+          // near future but it is still initialized for backward compatibility
+          geometryIds: [],
```
Comment (sedghi) [inline packages/tools/examples/videoContourSegmentation/index.ts:455]: You don't need this, you can just not provide data and it would work

File: packages/tools/examples/videoContourSegmentation/index.ts [hunk 15]
```diff
@@ -0,0 +1,484 @@
...
+    ]);
+
+  // Store the segmentation representation that was just created
+  segmentationRepresentationUID = segmentationRepresentationUIDs[0];
+
+  // Make the segmentation created as the active one
+  segmentation.activeSegmentation.setActiveSegmentationRepresentation(
+    toolGroupId,
+    segmentationRepresentationUID
+  );
```
Comment (sedghi) [inline packages/tools/examples/videoContourSegmentation/index.ts:477]: I would be pleasantly surprised if we remove this and it doesn't work, as it should become active internally. If it doesn't, then it might be a bug. We shouldn't have to set a segmentation to be active. Take a look at the labelmap examples, none of them have this logic.

File: packages/tools/examples/videoContourSegmentation/index.ts [hunk 16]
```diff
@@ -0,0 +1,484 @@
...
+
+  updateActiveSegmentIndex(1);
```
Comment (sedghi) [inline packages/tools/examples/videoContourSegmentation/index.ts:479]: same for this

File: packages/tools/src/eventListeners/mouse/mouseDownListener.ts [hunk 17]
```diff
@@ -153,6 +154,11 @@ function mouseDownListener(evt: MouseEvent) {
     evt.buttons === 1 ? DOUBLE_CLICK_TOLERANCE_MS : MULTI_BUTTON_TOLERANCE_MS
   );
 
+  if (evt.buttons >= MouseBindings.Fourth_Button) {
+    // Need to prevent the default handling so these bindings work, otherwise
+    // browsers will just navigate
```
Comment (sedghi) [inline packages/tools/src/eventListeners/mouse/mouseDownListener.ts:159]: I don't understand this comment, can you rephrase maybe?

**Review by sedghi — APPROVED** [review]
Thanks, looking good, see my remaining comments

File: packages/tools/examples/videoContourSegmentation/index.ts (hunk 15 above)
Comment (wayfarer3130) [reply packages/tools/examples/videoContourSegmentation/index.ts:477]: It is active internally, but it isn't associated with the toolGroup - I could manually add it, but am just going to leave this code for now.  i copied it direct from another example.

File: packages/tools/examples/videoContourSegmentation/index.ts [hunk 18]
```diff
@@ -0,0 +1,484 @@
...
+  const segmentationRepresentationUIDs =
+    await segmentation.addSegmentationRepresentations(toolGroupId, [
+      {
+        segmentationId,
+        type: csToolsEnums.SegmentationRepresentations.Contour,
+      },
+    ]);
+
+  // Store the segmentation representation that was just created
+  segmentationRepresentationUID = segmentationRepresentationUIDs[0];
```
[the remark this answers is not in the document]
Comment (wayfarer3130) [reply packages/tools/examples/videoContourSegmentation/index.ts:471]: No, because I'm assigning to a global variable, so that the representation can be used elsewhere.

File: packages/tools/examples/videoContourSegmentation/index.ts (hunk 13 above)
Comment (wayfarer3130) [reply packages/tools/examples/videoContourSegmentation/index.ts:444]: It is GUI logic, not part of the viewport itself, so I'm not sure where it would go if not here.  It really doesn't belong in CS3D, but if you have a good place to put all that logic, like maybe a helper method, I could do so.

File: packages/tools/examples/videoSplineROITools/index.ts (hunk 12 above)
Comment (wayfarer3130) [reply packages/tools/examples/videoSplineROITools/index.ts:288]: Already have the second one: https://www.cornerstonejs.org/docs/tutorials/basic-video
Adding the first one.

File: utils/demo/helpers/addManipulationBindings.ts (hunk 11 above)
Comment (wayfarer3130) [reply utils/demo/helpers/addManipulationBindings.ts:1]: ?? Not sure what you mean here.  The import is done in other helpers files, so I don't see the issue with this one either?

File: packages/tools/examples/videoContourSegmentation/index.ts (hunk 18 above)
[the remark this answers is not in the document]
Comment (sedghi) [reply packages/tools/examples/videoContourSegmentation/index.ts:466]: You could do ([segmentationRepresentationUID] = ... )

**Comment by sedghi** [comment]
Can you do me a favor and fix this also in this PR? should be very quick

CleanShot 2024-01-12 at 08 43 31

The rounding is too crazy

**Review by sedghi — APPROVED** [review]
